File: algorithms/N/Number of Sets of K Non-Overlapping Line Segments/Number of Sets of K Non-Overlapping Line Segments.js
var numberOfSets = function(n, k) {
    return combinations(n+k-1,2*k)%(1e9+7)
};
var combinations=(n,k)=>{
    var dp=[...Array(n+1)].map(d=>[...Array(k+1)].map(d=>1))
    for (let i = 1; i <=n; i++) 
        for (let k = 1; k <i; k++)
            dp[i][k]=(dp[i-1][k-1]+dp[i-1][k]) %(1e9+7)     
    return dp[n][k]
}

File: scripts/algorithms/N/Number of Sets of K Non-Overlapping Line Segments/Number of Sets of K NonOverlapping Line Segments.cpp
class Solution 
{
public:
    int mod=1e9+7;
    int dp[1011][1011][3];
    int f(int i, int k, int n, int ended)
    {
        if(k==0) return 1;
        if(i==n) return 0;
        if(dp[i][k][ended]!=-1) return dp[i][k][ended];
        int ans=0;
        if(ended) ans=(ans+f(i+1,k,n,0)+f(i+1,k,n,1))%mod;
        else ans=(ans+f(i+1,k,n,0)+f(i,k-1,n,1))%mod;
        return dp[i][k][ended]=ans;
    }
    int numberOfSets(int n, int k) 
    {
        memset(dp,-1,sizeof(dp));
        return f(0,k,n,1);
    }
};